src/JsonBuilder.cpp

```diff
@@ -6,25 +6,115 @@
 
+static unsigned
+Utf16ToUtf8(
+    _Out_writes_to_(cchSrc * 3, return) char unsigned* pchDest,
+    _In_reads_(cchSrc) char16_t const* pchSrc,
+    unsigned cchSrc)
+    noexcept
+{
+    unsigned iDest = 0;
+    for (unsigned iSrc = 0; iSrc != cchSrc; iSrc += 1)
+    {
+        auto const ch = pchSrc[iSrc];
+        if (ch < 0x80)
+        {
+            pchDest[iDest++] = static_cast<char unsigned>(ch);
+        }
+        else if (ch < 0x800)
+        {
+            pchDest[iDest++] = static_cast<char unsigned>(0xC0 | (ch >> 6));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | (ch & 0x3F));
+        }
+        else if (ch >= 0xD800 && ch < 0xDC00 &&
+            iSrc + 1 != cchSrc &&
+            pchSrc[iSrc + 1] >= 0xDC00 && pchSrc[iSrc + 1] < 0xE000)
+        {
+            // Surrogate pair.
+            iSrc += 1;
+            auto const ch32 = ((ch - 0xD800) << 10) + (pchSrc[iSrc] - 0xDC00) + 0x10000;
+            pchDest[iDest++] = static_cast<char unsigned>(0xF0 | (ch32 >> 18));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | ((ch32 >> 12) & 0x3F));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | ((ch32 >> 6) & 0x3F));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | (ch32 & 0x3F));
+        }
+        else
+        {
+            pchDest[iDest++] = static_cast<char unsigned>(0xE0 | (ch >> 12));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | ((ch >> 6) & 0x3F));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | (ch & 0x3F));
+        }
+    }
+
+    return iDest;
+}
+
+static unsigned
+Utf32ToUtf8(
+    _Out_writes_to_(cchSrc * 4, return) char unsigned* pchDest,
+    _In_reads_(cchSrc) char32_t const* pchSrc,
+    unsigned cchSrc)
+    noexcept
+{
+    unsigned iDest = 0;
+    for (unsigned iSrc = 0; iSrc != cchSrc; iSrc += 1)
+    {
+        auto const ch = pchSrc[iSrc];
+        if (ch < 0x80)
+        {
+            pchDest[iDest++] = static_cast<char unsigned>(ch);
+        }
+        else if (ch < 0x800)
+        {
+            pchDest[iDest++] = static_cast<char unsigned>(0xC0 | (ch >> 6));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | (ch & 0x3F));
+        }
+        else if (ch < 0x10000)
+        {
+            pchDest[iDest++] = static_cast<char unsigned>(0xE0 | (ch >> 12));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | ((ch >> 6) & 0x3F));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | (ch & 0x3F));
+        }
+        else
+        {
+            // If ch is outside the Unicode range, we'll ignore the high bits.
+            pchDest[iDest++] = static_cast<char unsigned>(0xF0 | (ch >> 18));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | ((ch >> 12) & 0x3F));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | ((ch >> 6) & 0x3F));
+            pchDest[iDest++] = static_cast<char unsigned>(0x80 | (ch & 0x3F));
+        }
```

Comment:
nit: some suggested comment here or for the function: 

- If the character is within the range 0x10000 to 0xFFFF, it is converted to a three-byte UTF-8 sequence.
- If the character is outside the Unicode Basic Multilingual Plane (BMP) (i.e., greater than or equal to 0x10000), it is converted to a four-byte UTF-8 sequence.

This might be over kill: 

- UTF-32 and UTF-8:
    - UTF-32 uses a fixed-width 32-bit value for each character, which means every character occupies exactly 4 bytes.
    - UTF-8, on the other hand, uses variable-length encoding, where characters can be represented using 1 to 4 bytes.
    - The function ensures that the UTF-32 characters are correctly mapped to their corresponding UTF-8 representations.